File: go/rz/redact.go
package rz

// PII redaction for log / exception-list views.
// The hash chain protects INTEGRITY, not CONFIDENTIALITY: audit entries carry
// customer IDs, invoice IDs, phone numbers and amounts in plaintext. This
// module provides a field-level redaction layer so any read surface can hide
// PII fields unless the caller is authorized. Redaction is intentionally
// conservative: it only masks at presentation time, never from the chain.

var PII_FIELDS = []string{"customerId", "customerName", "customerPhone", "customerEmail"}

const (
	PhoneLen  = 4
	EmailKeep = 2
)

func maskPhone(phone string) string {
	if len(phone) <= PhoneLen {
		return "••••"
	}
	return phone[:3] + "••••" + phone[len(phone)-2:]
}

func maskEmail(email string) string {
	at := indexOf(email, '@')
	if at <= 0 {
		return "•••@" + email[at+1:]
	}
	local := email[:at]
	domain := email[at+1:]
	shown := local
	if len(shown) > EmailKeep {
		shown = shown[:EmailKeep]
	}
	return shown + "•••@" + domain
}

func maskName(name string) string {
	if name == "" {
		return "•••"
	}
	parts := splitWords(name)
	if len(parts) <= 1 {
		return name
	}
	for i := range parts {
		if len(parts[i]) > 0 {
			parts[i] = string(parts[i][0]) + "•••"
		}
	}
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += " "
		}
		out += p
	}
	return out
}

// RedactPII returns a copy of record with the given PII fields masked.
func RedactPII(record map[string]any, fields ...string) map[string]any {
	if fields == nil {
		fields = PII_FIELDS
	}
	out := make(map[string]any, len(record))
	for k, v := range record {
		out[k] = v
	}
	for _, f := range fields {
		raw, ok := out[f]
		if !ok {
			continue
		}
		s, ok := raw.(string)
		if !ok {
			continue
		}
		switch f {
		case "customerPhone":
			out[f] = maskPhone(s)
		case "customerEmail":
			out[f] = maskEmail(s)
		case "customerName":
			out[f] = maskName(s)
		case "customerId":
			if len(s) > 4 {
				out[f] = "cust_•••" + s[len(s)-4:]
			} else {
				out[f] = "cust_•••" + s
			}
		}
	}
	return out
}

func indexOf(s string, b byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == b {
			return i
		}
	}
	return -1
}

func splitWords(s string) []string {
	var out []string
	start := -1
	for i := 0; i < len(s); i++ {
		if s[i] == ' ' || s[i] == '\t' {
			if start >= 0 {
				out = append(out, s[start:i])
				start = -1
			}
		} else {
			if start < 0 {
				start = i
			}
		}
	}
	if start >= 0 {
		out = append(out, s[start:])
	}
	return out
}

File: src/security/redact.ts
// PII redaction for log / exception-list views.
//
// The hash chain protects INTEGRITY, not CONFIDENTIALITY: audit entries carry
// customer IDs, invoice IDs, phone numbers and amounts in plaintext. Anyone with
// file access can read them. This module provides a field-level redaction layer
// so that any *read* surface (dashboard, exception list, auditor export) can
// hide PII fields unless the caller is authorized to see raw values.
//
// Redaction is intentionally conservative: it never reconstructs a masked value
// from the hash chain (that would defeat tamper-evidence), it only masks at
// presentation time.

export const PII_FIELDS = ['customerId', 'customerName', 'customerPhone', 'customerEmail'] as const;
export type PIIField = (typeof PII_FIELDS)[number];

const PHONE_LEN = 4;
const EMAIL_KEEP = 2;

function maskPhone(phone: string): string {
  return phone.length > PHONE_LEN ? phone.slice(0, 3) + '••••' + phone.slice(-2) : '••••';
}

function maskEmail(email: string): string {
  const at = email.indexOf('@');
  if (at <= 0) return '•••@' + email.slice(at + 1);
  const local = email.slice(0, at);
  const domain = email.slice(at + 1);
  const shown = local.slice(0, EMAIL_KEEP);
  return `${shown}•••@${domain}`;
}

function maskName(name: string): string {
  if (!name) return '•••';
  const parts = name.split(/\s+/);
  return parts.map((p) => (p.length > 0 && parts.length > 1 ? p[0] + '•••' : p)).join(' ');
}

export function redactPII<T>(record: T, fields: readonly PIIField[] = PII_FIELDS): T {
  if (!record || typeof record !== 'object') return record;
  const out: Record<string, unknown> = { ...(record as Record<string, unknown>) };
  for (const f of fields) {
    const raw = out[f];
    if (typeof raw !== 'string') continue;
    if (f === 'customerPhone') out[f] = maskPhone(raw);
    else if (f === 'customerEmail') out[f] = maskEmail(raw);
    else if (f === 'customerName') out[f] = maskName(raw);
    else if (f === 'customerId') out[f] = `cust_•••${raw.slice(-4)}`;
  }
  return out as T;
}

// Redact a full array of audit entries.
export function redactAuditEntries<T>(entries: T[]): T[] {
  return entries.map((e) => redactPII(e));
}
